pageserver/src/tenant/tasks.rs

```diff
@@ -207,6 +230,37 @@ async fn wait_for_active_tenant(
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+#[error("cancelled")]
+pub(crate) struct Cancelled;
+
+/// Provide a random delay for background task initialization.
+///
+/// This delay prevents a thundering herd of background tasks and will likely keep them running on
+/// different periods for more stable load.
+pub(crate) async fn random_init_delay(
+    period: Duration,
+    cancel: &CancellationToken,
+) -> Result<(), Cancelled> {
+    use rand::Rng;
+
+    let d = {
+        let mut rng = rand::thread_rng();
+
+        // gen_range asserts that the range cannot be empty, which it could be because period can
+        // be set to zero to disable gc or compaction, so lets set it to be at least 10s.
+        let period = std::cmp::max(period, Duration::from_secs(10));
+
+        // semi-ok default as the source of jitter
+        rng.gen_range(Duration::ZERO..=period)
+    };
+
+    tokio::select! {
+        _ = cancel.cancelled() => Err(Cancelled),
+        _ = tokio::time::sleep(d) => Ok(()),
```

Comment:
For the case where the configured period in the config is `Duration::ZERO`, the compaction and gc tasks have an effective period of 10 seconds, because they sleep 10 seconds.

So, instead of duplicating the 10 seconds constant here, you could re-order the callers to figure out what their effective period is first, then call this function with `NonZeroDuration`, if such a thing exists.

WDYT?

Reply:
I don't think a NonZeroDuration exists but could be made.

Reply:
Well, this would complicate the main loop of gc and compaction which I just got cleaned up from the mutation hell by splitting up configured period and sleep amount. I think this could be left undone or considered when the wait_for_active_tenant is removed with #3501.

Reply:
Also the 10 seconds might be a good harrison-stetson constant for just avoiding the thundering herd in general, with 1k+ tenants.

Reply:
Ok, let's go with this to not hold up the release for longer.